src/base/memory_monitor_sr.hpp

```diff
@@ -0,0 +1,231 @@
+      mmon_subcomponent &operator = (mmon_subcomponent &&) = delete;
+
+      /* function */
+      const char *get_name ();
+      void add_cur_stat (uint64_t size);
+      void sub_cur_stat (uint64_t size);
+
+    private:
+      std::string m_subcomp_name;
+      std::atomic<uint64_t> m_cur_stat;
+  };
+
+  class mmon_component
+  {
+    public:
+      mmon_component (const char *comp_name)
+	: m_comp_name (comp_name) {}
+      mmon_component (const mmon_component &) = delete;
+      mmon_component (mmon_component &&) = delete;
+
+      ~mmon_component () {}
+
+      mmon_component &operator = (const mmon_component &) = delete;
+      mmon_component &operator = (mmon_component &&) = delete;
+
+      /* function */
+      const char *get_name ();
+      void add_stat (uint64_t size, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int subcomp_idx, bool init);
+      int get_subcomp_index (const char *subcomp_name);
+      int add_subcomponent (const char *name);
+
+    private:
+      std::string m_comp_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_subcomponent>> m_subcomponent;
+  };
+
+  class mmon_module
+  {
+    public:
+      mmon_module (const char *module_name, const MMON_COMP_INFO *info);
+      mmon_module () {} /* for dummy */
+      mmon_module (const mmon_module &) = delete;
+      mmon_module (mmon_module &&) = delete;
+
+      virtual ~mmon_module () {}
+
+      mmon_module &operator = (const mmon_module &) = delete;
+      mmon_module &operator = (mmon_module &&) = delete;
+
+      /* function */
+      virtual int aggregate_stats (const MMON_MODULE_INFO &info);
+      void add_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init, bool expand);
+      void sub_stat (uint64_t size, int comp_idx, int subcomp_idx, bool init);
+      int add_component (const char *name);
+      inline int MAKE_STAT_IDX_MAP (int comp_idx, int subcomp_idx);
+
+    private:
+      /* const */
+      /* max index of component or subcomponent is 0x0000FFFF
+       * if some stats have max index of component or subcomponent,
+       * we don't have to increase it */
+      static constexpr int MAX_COMP_IDX = 0x0000FFFF;
+
+    private:
+      std::string m_module_name;
+      MMON_STAT m_stat;
+      std::vector<std::unique_ptr<mmon_component>> m_component;
+      std::vector<int> m_comp_idx_map;
+  };
+
+
+  class memory_monitor
+  {
+    public:
+      memory_monitor (const char *server_name)
+	: m_aggregater (this), m_server_name (server_name) {}
+      memory_monitor (const memory_monitor &) = delete;
+      memory_monitor (memory_monitor &&) = delete;
+
+      ~memory_monitor () {}
+
+      memory_monitor &operator = (const memory_monitor &) = delete;
+      memory_monitor &operator = (memory_monitor &&) = delete;
+
+      /* function */
+      int add_stat (THREAD_ENTRY *thread_p, MMON_STAT_ID stat_id, uint64_t size, bool expand);
```

Comment:
What is the purpose of the 'bool expand'?

Reply:
This is my reply of joohok's comment that has similar question about init and expand parameter. I think this reply can explain your question.

There were many aspects related to C++ that were not considered during the detailed design (e.g., the use of private) and as a result, there were changes in the design. It seems that there needs to be a discussion on whether to modify the document accordingly or simply leave it as a description in the story.

Due to the movement of member variables like vectors to the private section, in the top-level class (memory_monitor class), access to deeper levels and member variables of member variables became inaccessible. As a result, the structure was changed so that each class would request updates to their own member variables for the increment and decrement of stats.

The "init" and "expand" mentioned in the comments are additional parameters that arose from this process. They are necessary parameters to determine whether to increment init_stat along with cur_stat and whether to increase expand_count.

Since they are hidden from actual users, I think it's not a big problem.

Reply:
ex)
memmon_add_stat calls mmon_Gl->add_stat(thread_p, stat_id, size, false);
memmon_resize_stat calls mmon_Gl->resize_stat(thread_p, stat_id, old, new) and it will call mmon_Gl->add_stat(thread_p, stat_id, new, true);